### PR TITLE
fix m_vsnprintf return value

### PR DESCRIPTION
**Pull request by tius2000**
Changed the logic of m_vsnprintf() to return the correct value if the buffer is not large enough (see #1066 ). The overflow guard is no longer required. Unfortunately this required  some changes to the existing logic, I hope I did not break anything. Here is a minimal test application:

```
/*
	test return value of snprintf
	
	expected result:
	
		>123<      5
		>1234<     6
		>12345<    7
		>123456    8
		>123456    9
		% A 777 0x033A > 42< -1 > 3.14< >  123< >123  <
*/	

#include <SmingCore.h>

static void _test(const char *s) {
	char buf[8];
	size_t r = m_snprintf( buf, sizeof(buf), ">%s<", s );
	debugf("%-10.10s %d", buf, r);
}




void init() {
	Serial.begin(74880); 
	debugf("\n\ntest m_snprintf\n");
	
	//	test return values
	_test("123");
	_test("1234");
	_test("12345");
	_test("123456");
	_test("1234567");

	//	minimal format testing
	m_printf(
		"%% %c %o 0x%04X >%3u< %d >%5.2f< >%5.5s< >%-5.5s<\n", 
		65, 0777, 0x33A, 42, -1, 3.14159, "123", "123"
	);
}
```

### CODE REVIEW

**Comment by tius2000** [comment]
Well, C is not my primary language, I will give it a try to reduce the number of gotos ;-)

File: Sming/system/m_printf.cpp [hunk 1]
```diff
@@ -100,160 +94,134 @@ int m_vprintf ( const char * format, va_list arg )
...
+            case 'd':
+            case 'i':
+                s = ltoa_wp(va_arg(args, int), tempNum, base, width, pad);
+                break;
+
+            case 'f':
+                s = dtostrf_p(va_arg(args, double), width, precision, tempNum, pad);
+                break;
+
+            case 'o':
```
Comment (ADiea) [inline Sming/system/m_printf.cpp]: I would rewrite as
```
case 'o':
case 'x':
case 'X':
if('o' == f)
base=8;
else
base=16;
case 'u':

```

Comment (tius2000) [inline Sming/system/m_printf.cpp]: You are very strict regarding gotos ;-)

I changed it a bit to get rid of the fall-through, too.

**Comment by ADiea** [comment]
Have no other remarks. If "unit tests" in basic_string example work ok it's fine with me. Thanks again!

File: Sming/system/m_printf.cpp [hunk 2]
```diff
@@ -11,39 +11,33 @@ Descr: embedded very simple version of printf with float support
 
 #define MPRINTF_BUF_SIZE 256
 
-#define OVERFLOW_GUARD 24
-
 static void defaultPrintChar(uart_t *uart, char c) {
-	return uart_tx_one_char(c);
+    return uart_tx_one_char(c);
```
Comment (slaff) [inline Sming/system/m_printf.cpp]: The are multiple white-space changes that are not related to this PR. Please, revert them.

Comment (tius2000) [inline Sming/system/m_printf.cpp]: Sorry, if this is a FAQ - is there a guide regarding source code formatting for Sming (or even better, a command line tool that working with windows to do the correct formatting)? 

I tried to follow the rule 38 [here](http://geosoft.no/development/cppstyle.html#White%20Space): _Special characters like TAB and page break must be avoided.
These characters are bound to cause problem for editors, printers, terminal emulators or debuggers when used in a multi-programmer, multi-platform environment._ 

That's why I converted all tabs to space before committing.

Comment (slaff) [inline Sming/system/m_printf.cpp]: > That's why I converted all tabs to space before committing.

Changing of formatting and coding style can happen in another PR. This PR is only related to `m_vsnprintf` changes.

**Comment by tius2000** [comment]
* support for format and data strings in PROGMEM added
* format specifier '%S' added for compatibility (behaves like '%s)
* _m_vprintf_ now allocates less memory on the stack unless it is actually required

There is no longer a need for separate *_P methods, this should make it easier to solve the conflicts created by the macro _printf_P_ in FakePgmSpace.h.

File: Sming/system/m_printf.cpp [hunk 3]
```diff
@@ -68,27 +65,23 @@ int m_snprintf(char* buf, int length, const char *fmt, ...)
...
-		return 0;
-	}
-
-	char buf[MPRINTF_BUF_SIZE], *p;
+	size_t size = INITIAL_BUFFSIZE - 1;
+
+	//	need to retry if size is not big enough
+	while (1) {
+		char buffer[size + 1];
+		auto sz = m_vsnprintf(buffer, sizeof(buffer), fmt, va);
```
Comment (slaff) [inline Sming/system/m_printf.cpp]: Please, change `auto` to `size_t`.

File: Sming/system/m_printf.cpp [hunk 4]
```diff
@@ -68,27 +65,23 @@ int m_snprintf(char* buf, int length, const char *fmt, ...)
...
-	m_vsnprintf(buf, sizeof(buf), format, arg);
-
-	p = buf;
-	while (p && n < sizeof(buf) && *p)
-	{
-		cbc_printchar(cbc_printchar_uart, *p);
-		n++;
-		p++;
+		const char *p = buffer;
+		while (char c = *p++) cbc_printchar(cbc_printchar_uart, c);
```
Comment (slaff) [inline Sming/system/m_printf.cpp]: Format the while with opening and closing curly braces. 

```
while (char c = *p++)  {
     cbc_printchar(cbc_printchar_uart, c);
}
```

File: Sming/system/m_printf.cpp [hunk 5]
```diff
@@ -117,143 +110,122 @@ int m_printf(const char* fmt, ...)
...
-	}
-
-	*str = '\0';
-	return str - buf;
+    size_t size = 0;
+    auto add = [&](char c) {
+        if (++size < maxLen) *buf++ = c;
+    };
+
+    while ( char f = pgm_read_byte(fmt) ) {
```
Comment (slaff) [inline Sming/system/m_printf.cpp]: If you read every byte using `pgm_read_byte` that will cause overhead. Is that really needed in a very core component? What are the advantages?

Comment (tius2000) [inline Sming/system/m_printf.cpp]: You a right, this will make every call to *printf a tiny bit slower. The code size is almost identical. I see the following advantages:

- it will reduce total flash size and complexity as a many additional methods are no longer needed
- all classes and methods using _m_printf_ will benefit immediately
- life will be much easier for users as they do not need to care which method to call and which format specifier to use
- format strings should go to flashmem anyway to reduce ram size, it is faster than the existing implementation for format strings in flashmem
- printf is mostly used for serial output where execution speed does not matter that much
- for time critical operations direct c string manipulation should be used instead of the printf functions

IMHO the main advantage is that it eliminated a major pitfall for users working with progmem.

**Comment by tius2000** [comment]
I'm quite new to GitHub - how can I create a PR that depends on another PR?

**Comment by slaff** [comment]
> I'm quite new to GitHub - how can I create a PR that depends on another PR?

Create a branch for PR 1. Work on it until you are ready. Then create another branch for PR1 that is based on PR1.
```
git check PR1
# ... work...
# .... ready
git branch PR2
git checkout PR2
# additional work 
```

If there are changes in PR1 you can keep PR2 up to date using the following commands:

```
git checkout PR1 
# .. make sure that you have the latest code for PR1
git checkout PR2
git merge PR1
# git mergetool # if needed
git rebase PR1
# git mergetool # if needed
git push -f 
``` 

That should be it.

**Comment by tius2000** [comment]
I created a new PR #1096 that way, hope I got it right!
